sympy/ntheory/residue_ntheory.py

```diff
@@ -1300,3 +1300,40 @@ def discrete_log(n, a, b, order=None, prime_order=None):
         return _discrete_log_pollard_rho(n, a, b, order)
 
     return _discrete_log_pohlig_hellman(n, a, b, order)
+
+
+def collatz(n):
+    """Starting with any positive integer N, Collatz sequence corresponding
+    to N as the numbers formed by the following operations:
+    N → N/2 ( if N is even)
+    N → 3N + 1 (if N is odd)
+    It is conjectured but not yet proven that no matter which
+    positive integer we start with; we always end up with 1.
+
+    Examples
+    ========
+
+    >>> from sympy.ntheory.residue_ntheory import collatz
+    >>> list(collatz(10))
+    [10, 5, 16, 8, 4, 2, 1]
+    >>> list(collatz(15))
+    [15, 46, 23, 70, 35, 106, 53, 160, 80, 40, 20, 10, 5, 16, 8, 4, 2, 1]
+    >>> list(collatz(21))
+    [21, 64, 32, 16, 8, 4, 2, 1]
+
+    References
+    ==========
+
+    .. [1] http://mathworld.wolfram.com/CollatzProblem.html
+    """
+
+    n = as_int(n)
```

Comment:
It need not be an integer literal
```python
>>> def collatz(n):
...   if not n.is_integer:
...       raise ValueError
...   while True:
...     yield n
...     last = n
...     if n % 2 == 0:
...         n = n/2
...     elif n % 2 == 1:
...         n = 3*n + 1
...     if n in (1, last):
...         break
...
...
>>>
>>> list(collatz(12*Symbol('n',integer=True)+4))
[12*n + 4, 6*n + 2, 3*n + 1]
```
The docstring could indicate that the series terminates when the parity of the expression cannot be determined. One could separate literal and non-literal handling, perhaps.

Reply:
I would rather raise an exception when the parity cannot be determined. You can sympify the input and check if `is_even` is True or False, and raise an exception on None.

Reply:
> raise an exception when the parity cannot be determined

The code I suggested would just return the value unchanged. So 1 or an indeterminate parity expression would be returned as the only value.

Reply:
As Collatz sequence is a sequence up to 1. I don` think that it is a good idea to break when we can`t decide parity. I don`t think that someone will need a collatz sequence of variable. We could just check if it is a variable, then we can raise the required exception. But if you think so, I will commit the changes. Please let me know.

Reply:
> I don`t think that someone will need a collatz sequence of variable

With a variable you can show that regardless of the exact starting value, values that *end* in a certain way will merge with each other:
```python
>>> list(s(32*i+16+7))
[32*i + 23, 96*i + 70, 48*i + 35, 144*i + 106, 72*i + 53, 216*i + 160, 108*i + 80, 54*i + 40,
27*i + 20]
>>> list(s(16*i+8+3))
[16*i + 11, 48*i + 34, 24*i + 17, 72*i + 52, 36*i + 26, 18*i + 13, 54*i + 40, 
27*i + 20]
```
So that shows that all numbers that end with `10111` in binary will merge with all those that end with `1011`. That's similar to the relationships where an even will merge with its half, or that an odd number will merge with a `number = 4*odd + 1`
```python
>>> n=2*i+1
>>> list(s(n))
[2*i + 1, 6*i + 4, 3*i + 2]
>>> list(s(2*n))
[4*i + 2, 2*i + 1, 6*i + 4, 3*i + 2]
>>> list(s(4*n+1))
[8*i + 5, 24*i + 16, 12*i + 8, 6*i + 4, 3*i + 2]
```
The 'Sym' of SymPy is for 'symbolic' ;-)

Reply:
Actually instead of an exception it could yield the Piecewise form.

Reply:
```python
if n%2 not in (0, 1):
    yield Piecewise((3*n + 1, Eq(Mod(n, 2), 1)), (n/2, True))
else:
    # current routine
```